saga/seata-saga-engine/src/main/java/io/seata/saga/engine/pcext/handlers/ScriptTaskStateHandler.java

```diff
@@ -0,0 +1,164 @@
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * ScriptTaskState Handler
+ *
+ * @author lorne.cl
+ */
+public class ScriptTaskStateHandler implements StateHandler, InterceptableStateHandler {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ScriptTaskStateHandler.class);
+
+    private List<StateHandlerInterceptor> interceptors = new ArrayList<>();
+
+    private volatile Map<String, ScriptEngine> scriptEngineCache = new ConcurrentHashMap<>();
+
+    @Override
+    public void process(ProcessContext context) throws EngineExecutionException {
+
+        StateInstruction instruction = context.getInstruction(StateInstruction.class);
+        ScriptTaskStateImpl state = (ScriptTaskStateImpl) instruction.getState(context);
+
+        String scriptType = state.getScriptType();
+        String scriptContent = state.getScriptContent();
+
+        Object result;
+        try {
+
+            List<Object> input = (List<Object>) context.getVariable(DomainConstants.VAR_NAME_INPUT_PARAMS);
+
+            if (LOGGER.isDebugEnabled()) {
+                LOGGER.debug(">>>>>>>>>>>>>>>>>>>>>> Start to execute ScriptTaskState[{}], ScriptType[{}], Input:{}",
+                        state.getName(), scriptType, input);
+            }
+
+            StateMachineConfig stateMachineConfig = (StateMachineConfig) context.getVariable(
+                    DomainConstants.VAR_NAME_STATEMACHINE_CONFIG);
+
+            ScriptEngine scriptEngine = getScriptEngineFromCache(scriptType, stateMachineConfig.getScriptEngineManager());
+            if (scriptEngine == null) {
+                throw new EngineExecutionException("No such ScriptType[" + scriptType + "]",
+                        FrameworkErrorCode.ObjectNotExists);
+            }
+
+            Bindings bindings = null;
+            Map<String, Object> inputMap = null;
+            if (input != null && input.size() > 0 && input.get(0) instanceof Map) {
+                inputMap = (Map<String, Object>) input.get(0);
+            }
+            List<Object> inputExps = state.getInput();
+            if (inputExps != null && inputExps.size() > 0 && inputExps.get(0) instanceof Map) {
+                Map<String, Object> inputExpMap = (Map<String, Object>) inputExps.get(0);
+                if (inputExpMap.size() > 0) {
+                    bindings = new SimpleBindings();
+                    for (String property : inputExpMap.keySet()) {
+                        if (inputMap.containsKey(property)) {
+                            bindings.put(property, inputMap.get(property));
+                        }
+                        else {
+                            //if we do not bind the null value property, groovy will throw MissingPropertyException
+                            bindings.put(property, null);
+                        }
+                    }
+                }
+            }
+            if (bindings != null) {
+                result = scriptEngine.eval(scriptContent, bindings);
```

Comment:
Will this script be parsed every time？

Reply:
I have cached ScripteEngine for each script type, and ScripteEngine will cache compiled scripts by default.

Reply:
OK.